### PR TITLE
Add asserts in haveDifferentSharding

### PR DESCRIPTION
**Pull request by samnordmann**
Adresses issue https://github.com/NVIDIA/Fuser/issues/1656

### CODE REVIEW

**Review by naoyam — APPROVED** [review]
LGTM

File: csrc/multidevice/utils.cpp [hunk 1]
```diff
@@ -60,6 +61,9 @@ std::unordered_set<TensorView*> haveDifferentSharding(
   }
 
   for (auto tv : tvs) {
+    // isSharded asserts that there are no split/merge and that only the outmost
```
Comment (naoyam) [inline csrc/multidevice/utils.cpp:64]: Actually, let's move the assumptions to the beginning of this function as it also affects how we can use `ComputeAtMap`.

Comment (naoyam) [reply csrc/multidevice/utils.cpp:64]: Also, shouldn't we check `ref` as well?

**Comment by samnordmann** [comment]
!build

**Pull request merged**